Resistance

hopshackle opened the pull request:
Implementation of Resistance 

Includes some additions to AbstractGameState to support games with 'teams', in which all members of a team win/lose together. RunGames and ParameterSearch will now detect this, and act appropriately as default (i.e. all members of the same team will get a copy of the same agent type).

rdgain commented on src/main/java/games/resistance/ResGameState.java:
```diff
@@ -0,0 +1,283 @@
+        MissionVote, TeamSelectionVote, LeaderSelectsTeam
+    }
+
+    List<PartialObservableDeck<ResPlayerCards>> playerHandCards = new ArrayList<>(10);
```
I think this is better separated, with 1 ID/alliance flag for each player, and one voting array. I'm not sure if it needs to go through a PO deck with 3 cards per player with specific access at idx 2 to get a player's identity for example, it feels overcomplicated. Isn't the voting in a separate array already anyway? 

Then the card extending class could just be an enum, since it doesn't really have any other role?

hopshackle replied:
I agree it's sub-optimal (the use of Cards), but I don;t want to change it as it's also deeply entwined in the Resistance GUI. (And it works.)

However, voting now refactored as you suggest to a single mechanism.

rdgain reviewed (changes requested):
+ 1 other comment: some actions are missing toString implementations :)

hopshackle commented on src/main/java/games/resistance/ResGameState.java:
```diff
@@ -0,0 +1,283 @@
+    ResPlayerCards.CardType[] missionVotingChoice;
+
+    List<Integer> teamChoice;
+    IGamePhase previousGamePhase = null;
```
Fixed.

hopshackle commented on src/main/java/games/resistance/ResParameters.java:
```diff
@@ -0,0 +1,74 @@
+
+import java.util.*;
+
+public class ResParameters extends AbstractParameters {
```
Fixed

hopshackle replied:
Tidied up a but to make it more accessible for future work on changing the numbers of spies etc.
(But turning this into something full Tunable is future work.)

hopshackle commented:
I've addressed all these points, with the major execption of the use of Cards instead of simple enums - which I fully agree would be the correct way to do it, but changing it at this stage feels like to much work (and risk) for little gain.

rdgain reviewed (approved):
I think it also needs an update from master, but approved!
